Sousa+18 parameter update broke author/links

The newest parameters are not deployed yet.

The build fails with 
```
__________________________ test_homepage_status_code ___________________________
client = <FlaskClient <Flask 'app'>>
    def test_homepage_status_code(client):
>       homepage = client.get(url_for("homepage"))
tests/test_app.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
...
app.py:16: in homepage
    df['alink'] = list(map(author_html, df['Author'], df['link']))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
author = nan, link = nan
    def author_html(author, link):
        """
        Create HTML anchor tag for author with correct link to ADSABS
    
        Inputs
        ------
        author : str
          Name of the author(s)
        link : str
          Link to article
    
        Output
        ------
        alink : str
          Author anchor tag to link
        """
>       if (',' in author) and (',' in link):
E       TypeError: argument of type 'float' is not iterable
utils.py:313: TypeError
```
The author and link are both `nan` so the code fails.

Two possible options:
-  We need to add a patch to allow author/links to be missing. Change the NULL to the required null string early. The input types for `author_html() ` are already specified as string so the fix should probably go outside this function so that it does not get given the float `nan`.

OR

- There is an problem with the database itself that needs to be fixed. 
    @acsferreira Are there any authors or links that are supposed to be NULL?

--- a/sweetercat/tests/test_data.py
+++ b/sweetercat/tests/test_data.py
@@ -1,4 +1,6 @@
 # Test properties of the sweet-cat database
+import pytest
+
 
 def test_all_have_star_name(SCdata):
     # Test every entry has a "Star" name
@@ -10,11 +12,25 @@ def test_all_have_link(SCdata):
     # Test each entry has an associated author/paper.
     df, _ = SCdata
     null_links = df.link.isnull()
-    # Printing is for idenifying missing links on failure only
+    # printing is for identifying missing links on failure only
     print("Stars with missing links:\n")
     for star in df.Star[null_links].values:
         print(star)
     assert not null_links.values.any()
+
+
+# xfail on purpose (to indicate there are missing links only)
+@pytest.mark.xfail
+def test_links_are_not_self_generated(SCdata):
+    # Test if any have exoplanet.eu link from generate_missing_links.
+    df, _ = SCdata
+    generated_links = df.link.str.contains("http://exoplanet.eu/catalog/")
+    print(generated_links)
+    # printing is for identifying generated_links on failure only
+    print("Stars with artificial links:\n")
+    for star in df.Star[generated_links].values:
+        print(star)
+    assert not generated_links.values.any()
 
 
 def test_all_have_Author(SCdata):
@@ -22,7 +38,7 @@ def test_all_have_Author(SCdata):
     df, _ = SCdata
 
     null_author = df.Author.isnull()
-    # Printing is for idenifying missing author on failure only.
+    # printing is for identifying missing author on failure only.
     print("Stars with missing Author:")
     for star in df.Star[null_author].values:
         print(star)

--- a/sweetercat/tests/test_utils.py
+++ b/sweetercat/tests/test_utils.py
@@ -8,7 +8,7 @@ import pytest
 
 from utils import absolute_magnitude, hz, planetAndStar, plDensity, readSC
 from utils import table_convert, stellar_radius, planetary_radius, get_default
-from utils import luminosity, author_html
+from utils import luminosity, author_html, generate_missing_link
 
 
 def test_absolute_magnitude():
@@ -161,3 +161,16 @@ def test_multiple_author_html():
     alink = author_html(author, link)
     expected = '<a target="_blank" href="hobbiton.me">Frodo Baggins</a>, <a target="_blank" href="greendragon.io">Samwise Gamgee</a>'
     assert alink == expected
+
+
+@pytest.mark.parametrize(
+    "input,expected",
+    [
+        (None, "http://exoplanet.eu/catalog/"),
+        ("CoRoT-31", "http://exoplanet.eu/catalog/corot-31_b/"),
+        ("gamma 1 Leo", "http://exoplanet.eu/catalog/gamma_1_leo_b/"),
+        ("SDSS1411+2009", "http://exoplanet.eu/catalog/sdss1411+2009_b/"),
+    ],
+)
+def test_generate_missing_link(input, expected):
+    assert generate_missing_link(input) == expected

--- a/sweetercat/utils.py
+++ b/sweetercat/utils.py
@@ -343,8 +343,8 @@ def generate_missing_link(star=None):
         Link to exoplanet.eu.
     """
     if star is None:
-        link = "http://exoplanet.eu/catalog/"
+         return "http://exoplanet.eu/catalog/"
     else:
-        star_eu = star.lower().replace(" ", "_")
-        link = "http://exoplanet.eu/catalog/{}_b/".format(star_eu)
-    return link
+        star_formatted = star.lower().replace(" ", "_")
+        return "http://exoplanet.eu/catalog/{}_b/".format(star_formatted)
+

--- a/sweetercat/utils.py
+++ b/sweetercat/utils.py
@@ -148,11 +148,16 @@ def readSC(nrows=None):
         df['lum'] = list(map(luminosity, df['teff'], df['Vmag'], df['par']*1E-3, df['mass']))
         df['Star'] = df['Star'].str.strip()
 
+        # Generate missing link https://github.com/DanielAndreasen/SWEETer-Cat/issues/135
+        for star in df.Star[df['link'].isnull()].values:
+            df.loc[df['Star'] == star, ['link']] = generate_missing_link(star)
+
         plots = ['Vmag', 'Vmagerr', 'Vabs', 'par', 'parerr', 'teff', 'tefferr',
                  'logg', 'loggerr', 'logglc', 'logglcerr', 'vt', 'vterr',
                  'feh', 'feherr', 'mass', 'masserr', 'lum']
         cache.set('starDB', df, timeout=5*60)
         cache.set('starCols', plots, timeout=5*60)
+
     if nrows is not None:
         return df.loc[:nrows-1, :], plots
     return df, plots
@@ -320,3 +325,26 @@ def author_html(author, link):
     else:
         alink = '<a target="_blank" href="{}">{}</a>'.format(link, author)
     return alink
+
+
+def generate_missing_link(star=None):
+    """Give exoplanet.eu link if link is missing.
+
+    See https://github.com/DanielAndreasen/SWEETer-Cat/issues/135.
+
+    Parameters
+    ----------
+    Star: str Optional
+        Star name from Sweet-cat.
+
+    Returns
+    -------
+    link: str
+        Link to exoplanet.eu.
+    """
+    if star is None:
+        link = "http://exoplanet.eu/catalog/"
+    else:
+        star_eu = star.lower().replace(" ", "_")
+        link = "http://exoplanet.eu/catalog/{}_b/".format(star_eu)
+    return link

